entropy: Bluetooth HCI entropy source added

JordanYates opened the pull request:
Implements an entropy source that utilizes the Bluetooth HCI command `LE_RAND` as its randomness source.

Changes to the xoshiro128 driver are required as the HCI interface is not available until after the application calls `bt_enable`, which is unlikely to be before `PRE_KERNEL_2`, where the random drivers are initialized.

Implements #37186

Commit:
Bluetooth: check if Bluetooth has been enabled

Adds a simple boolean check of whether the Bluetooth subsystem has been
enabled. This allows users outside of `subsys/bluetooth` to check
whether they can send HCI commands.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
random: xoshiro128: defer state init

Defer the state initialization of the xoshiro algorithm until the first
time a random number is requested. This allows the PRNG algorithm to be
used with entropy sources which may not be available at boot, i.e
Bluetooth HCI.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
entropy: bt_hci: driver added

Adds an entropy driver that uses Bluetooth HCI commands as its source
of randomness. As this method is blocking, the ISR API is not supported.

As this method will range from relatively slow (same core Bluetooth HCI
controller) to extremely slow (UART HCI Bluetooth controller), use the
xoshiro PRNG by default for RNG generation.

Implements #37186

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
random: xoshiro128: reseed state on entropy fail

When requesting seed entropy data fails, reseed the PRNG state with
runtime data. This is a minimal effort to ensure that random data
requested before a backing entropy device is ready does not result in
repeatable data on each boot.

The random XOR integers are selected from the CRC32 algorithm.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
tests: entropy: test entropy_bt_hci

Add a testcase to validate behaviour of the ENTROPY_BT_HCI driver.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
boards: nrf5340dk_nrf5340: Bluetooth RNG

Use the ENTROPY_BT_HCI entropy device on the application core, which has
no access to the hardware RNG peripheral (network core only).

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

jhedberg commented:
What are the implications of this to the `bt_rand()` API? Not related, or could that API now be dropped in favor of using some system-wide random number generation API?

jhedberg commented:
> What are the implications of this to the bt_rand() API? Not related, or could that API now be dropped in favor of using some system-wide random number generation API?

In particular, the pure Bluetooth host version of `bt_rand()` as implemented in https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/crypto.c is kind of similar, in that it uses this same HCI command to do seeding, and then uses a TinyCrypt PRNG to generate random data.

dleach02 commented:
> What are the implications of this to the `bt_rand()` API? Not related, or could that API now be dropped in favor of using some system-wide random number generation API?

It does seem that, at a minimum, bt_rand() could just redirect to the system rand(). But this is were you need to determine if the BT rand values are cryptographic values or just random numbers to determine which rand to connect with

dleach02 commented on boards/arm/nrf5340dk_nrf5340/nrf5340_cpuapp_common.dts:
```diff
@@ -12,6 +12,7 @@
 		zephyr,uart-mcumgr = &uart0;
 		zephyr,bt-mon-uart = &uart0;
 		zephyr,bt-c2h-uart = &uart0;
+		zephyr,entropy = &rng_hci;
```
does the nrf5340 not have its own entropy generation source?

JordanYates replied:
Not accessible to the application core. The `RNG` peripheral is only available to the network core, which runs the Bluetooth controller. AFAIK right now there is no mechanism to get random numbers on the application core, hence this PR.

tejlmand replied:
well, the nRF5340 appcore has access to cryptocell and its RNG, which can be chosen as:
```
zephyr,entropy = &cryptocell;
```

@frkv @mbolivar-nordic could one of you take a look here ?

frkv replied:
Follow up with a little bit more information for this

The app-core has a special relationship to the CryptoCell or any local peripherals on the device. This is built into the mechanism of SPU and the addition of TrustZone security extensions. There is much more security in using this peripheral as-is

There is no need to use IPC communication for the RNG which is much more distanced from security processes and the "Root of Trust" established during boot.

The RNG on the netcore is intended to be used on the radio core, only. What you are providing by exporting this functionality from the netcore to the app-core is an anti-pattern which is not part of our HW design.

dleach02 commented:
Why is only the xoshiro128 driver impacted by this? Can you not use this entropy source for rand32_ctr_drbg.c (assuming the entropy generated is cryptographically secure)? 

It actually presents problems for the ctr because you can't just give that rand function some static values while you wait on the entropy source to be ready.

JordanYates commented:
>> What are the implications of this to the bt_rand() API? Not related, or could that API now be dropped in favor of using some system-wide random number generation API?

> It does seem that, at a minimum, bt_rand() could just redirect to the system rand(). But this is were you need to determine if the BT rand values are cryptographic values or just random numbers to determine which rand to connect with

`bt_rand` has the advantage of a fixed PRNG implementation, so the Bluetooth stack can be sure that (most of the time) the calls will be fast, which is not necessarily the case with the system RNG. I don't know whether fast RNG is required for the correct operation of the host however.

JordanYates commented:
> In particular, the pure Bluetooth host version of bt_rand() as implemented in https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/crypto.c is kind of similar, in that it uses this same HCI command to do seeding, and then uses a TinyCrypt PRNG to generate random data.

That implementation was the inspiration for this driver. It cannot be used directly as an entropy device as it is already a PRNG output.

dleach02 commented:
> 
> `bt_rand` has the advantage of a fixed PRNG implementation, so the Bluetooth stack can be sure that (most of the time) the calls will be fast, which is not necessarily the case with the system RNG. I don't know whether fast RNG is required for the correct operation of the host however.

Some implementations use the HW entropy device as their RNG source and when entropy is exhausted it has to regenerate more entropy... which can take a bit of time. On those platforms, we have tied the entropy to provide seed to Xoro and the ctr algorithms which have a fairly constant rate of supplying RNG values.

But I'm still wondering if this target platform ever needs cryptographically secure RNG values?

JordanYates commented:
> Why is only the xoshiro128 driver impacted by this? Can you not use this entropy source for rand32_ctr_drbg.c (assuming the entropy generated is cryptographically secure)?

AFAICT this can "almost" be used with `rand32_ctr_drbg` as is. `ctr_drbg` is only initialized when it is first called, so it's behavior is the same as the updated `xoshiro`. It doesn't appear to handle failed initialization properly, but I will add a commit to fix that.

Commit:
random: ctr_drbg: initialize on first run

With f44b3dc4dfef4ba0adde064e910f35521f09232f statically initializing
the entropy_dev variable, `ctr_drbg_initialize` was not being run on
the first call to `z_impl_sys_csrand_get`. Use a dedicated static bool
to track whether the init needs to be run.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

carlescufi commented on drivers/entropy/entropy_bt_hci.c:
```diff
@@ -0,0 +1,57 @@
+	struct bt_hci_rp_le_rand *rp;
+	struct net_buf *rsp;
+	int req, ret;
+
```
Please check if you are in ISR, since you cannot send HCI commands from interrupt context.

JordanYates replied:
Is this required given the existence of the `get_entropy_isr` API function, which this driver does not implement?

carlescufi replied:
That's fair, let's leave it here.

mbolivar-nordic commented on drivers/entropy/entropy_bt_hci.c:
```diff
@@ -0,0 +1,57 @@
+	.get_entropy_isr = NULL
+};
+
+DEVICE_DT_INST_DEFINE(0, entropy_bt_init,
```
I think this driver needs to be single-instance because the BT HCI API assumes a singleton instance, right?

I still think it makes sense to use DT _INST_FOREACH_STATUS_OKAY to allow users to `status = "disabled";` the node when hacking around to prevent device allocation.

JordanYates replied:
done

JordanYates commented:
I still need some input on how to configure the tests to run correctly. Is it even possible to test again a HCI controller in CI?

mbolivar-nordic reviewed (commented):
DT portions LGTM

tejlmand reviewed (dismissed):
lgtm

mbolivar-nordic commented:
> I still need some input on how to configure the tests to run correctly. Is it even possible to test again a HCI controller in CI?

@carlescufi @aescolar any input here?

aescolar commented:
> I still need some input on how to configure the tests to run correctly. Is it even possible to test again a HCI controller in CI?

Yes. The nrf52_bsim can be compiled with the bt controller. If you do not need any BT peers, it should be practically as easy as other ztests (if you don't need a BT network you can just run it detached from a babblesim simulation, it will look very similar to a native_posix run from the usage point of view).
Take a look here: https://docs.zephyrproject.org/2.7.0/boards/posix/nrf52_bsim/doc/index.html .

mbolivar-nordic commented:
@JordanYates can you address the compliance failure?

tejlmand reviewed (dismissed):
lgtm - reapproved

mbolivar-nordic commented:
@JordanYates I'm still seeing compliance failures

JordanYates commented:
> @JordanYates I'm still seeing compliance failures

That was unrelated to this PR (fixed in #43355). Hopefully the rebase fixes it

tejlmand reviewed (dismissed):
lgtm - reapproved

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3627,6 +3627,11 @@ int bt_enable(bt_ready_cb_t cb)
 
+bool bt_is_ready(void)
+{
+	return atomic_test_bit(bt_dev.flags, BT_DEV_READY);
```
Is this really the right flag to check for and not BT_DEV_ENABLE? The `bt_disable()` API being added in #42409 doesn't seem to touch BT_DEV_READY at all. (ping @erwango )

JordanYates replied:
`BT_DEV_ENABLE` is set as practically the first action of `bt_enable`. The Bluetooth system is not ready to perform any actions at that point. My view is that external users want to check if they can actually use Bluetooth, not if someone has called `bt_enable`.

carlescufi commented:
> > I still need some input on how to configure the tests to run correctly. Is it even possible to test again a HCI controller in CI?
> 
> @carlescufi @aescolar any input here?

No, unfortunately not easily unless we use native_posix + bsim. But I think that's overkill right here.

carlescufi commented:
@JordanYates  could you please rebase this?

Thalley commented:
> @JordanYates could you please rebase this?

Why do you want a rebase? It doesn't seem to have any conflicts, and if you want to run the tests on newest main, the tests should be rerun instead. Rebasing will risk dismissing all the approvals :(

carlescufi commented:
@JordanYates this is failing CI
